### ISSUE TITLE
Deal with disconnected cDBGs

### ISSUE THREAD

**Issue by bdsullivan** [spacegraphcats/spacegraphcats#97]
It seems like datasets with large numbers of connected components could be causing massive slowdowns in our code -- both because of the I/O for checkpointing, and also because our "early out" for stopping refinement of the domgraph when it gets "small" fails.

Needs some thought.

**Comment by bdsullivan** [comment]
How are we ended up with more vertices in our top-level domgraph than components in the original dataset (podar) ?

**Comment by bdsullivan** [comment]
Also, hypothesis for excessive repeated refinement is having something analogous to long-legged spiders in domgraph.

**Comment by microgravitas** [comment]
More vertices in domgraph than components points to an issue with how the domgraph edges are computed

**Comment by sgsmob** [comment]
Felix is correct.  When I added an assertion error that the dominators of each vertex actually form a clique, it failed.  Probably won't have time to fully diagnose this before my next flight but may get to it this afternoon after I sleep some.

**Comment by sgsmob** [comment]
After some digging, I believe I've found the problem.  I still believe the math is correct in that one only needs to consider the "boundary" vertices of each dominator `x` (those vertices that are neighbors of a vertex dominated by `x` but are not themselves dominated by `x`) to determine the edge structure of `x`.  However, since our graph by the time we compute the domset has unidirectional edges, it is not possible to compute the boundary vertices in the obvious/straightforward way.  I suspect it is not even possible to do efficiently at all given the data structures.  Not sure what better method we can use for domgraph edge computation, but I will think some more about it.

Bottom line, though, is that the currently computed catlases are definitely not correct.

**Comment by sgsmob** [comment]
I did the most basic thing to fix this issue, which is to make all the weight 1 arcs bidirectional (i.e. re-add them to the graph) before computing the dominating sets.  From a principled stance, I'm really unsatisfied by this fix but it appears to work pretty well.  I was able to get a podar catlas made in less than 20 minute, which included the checkpointing.  I'm re-running now without checkpointing to see how long it takes.

**Comment by sgsmob** [comment]
Without the checkpoints it takes 14 minutes, not bad!

**Comment by sgsmob** [comment]
I'm going to close this issue because the immediate problem of the domgraphs getting disconnected is no longer an issue.
